app/src/main/java/org/simple/clinic/analytics/TimingCollector.kt

```diff
@@ -0,0 +1,98 @@
+package org.simple.clinic.analytics
+
+import androidx.annotation.VisibleForTesting
+import org.simple.clinic.util.UtcClock
+import org.threeten.bp.Duration
+import org.threeten.bp.Instant
+import java.util.UUID
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * This class is meant to collect timing information for an operation over time. An operation can
+ * have multiple stages, and the collector allows collecting discrete timing information for each
+ * of these stages.
+ *
+ * Every instance of the collector is given a run ID, so that all events from a given run can be
+ * identified and grouped together if required.
+ *
+ * ### Example Usage
+ * ```
+ * val collector = TimingCollector("Search for patients", utcClock)
+ *
+ * collector.start("fetch-patient-names")
+ * // Fetch patient names from database
+ * collector.stop("fetch-patient-names")
+ *
+ * collector.start("fuzzy-name-matching")
+ * // Use fuzzy name matching to drop patients that don't match  criteria
+ * collector.stop("fuzzy-name-matching")
+ *
+ * collector.start("fetch-patient-details")
+ * // Fetch complete patient details from database
+ * collector.stop("fetch-patient-details")
+ *
+ * collector.start("sort-by-facility")
+ * // Sort the final search results list according the patients visited facility list
+ * collector.stop("stop-by-facility")
+ *
+ * // Once the entire operation is complete, push higher level events to analytics
+ * collector.finish()
+ * ```
+ *
+ * **Note:** Since an operation can happen across multiple threads, this class has been designed
+ * with thread-safety in mind. So it uses immutable maps for tracking operations, and atomic
+ * primitives where necessary.
+ **/
+class TimingCollector(
+    val name: String,
```

Comment:
I am not sure what this `name` means?
Is it the `operationName` or `eventName`? 
Can we call it that if so?

Reply:
I also want to do this to have a clear indication of what an `operation` is and what a `stage` is in context to this class.